fix: Intent handling logic fixes

janhartman opened the pull request:
I was looking into why chat/executed events in telemetry had 0 intent corrections. It turns out that there's a bit of buggy logic that was preventing telemetry from working as intended.

Two fixes:
- Intent prefetching didn't run if the query had no @-mentions . Now, it does and the experience is quite a bit snappier.
- We now properly track the user's selected intent so we can log it alongside the model-detected intent.

It would also be great to land https://github.com/sourcegraph/cody/pull/6572 alongside this one - what's the status there @thenamankumar ?

There's still one thing bothering me - the prefetching is too eager because it's based on editor state, meaning if you just click around, select text but make no edits, it'll still trigger. We should fix that! [Here's a Loom showing the problem](https://www.loom.com/share/f8790d65bcfd4073b04e6e1a5fa0554d).

## Test plan
Verify that intent detection still works correctly.
Apart from that, verify that in the `chat/executed` telemetry event, we now receive different `detectedIntent` and `userSpecifiedIntent` values when an intent correction is made.

Unblocks [SPLF-330](https://linear.app/sourcegraph/issue/SPLF-330/make-sure-the-collected-telemetry-is-usable-for-fine-tuning-the-model).

janhartman commented on vscode/webviews/chat/Transcript.tsx:
```diff
@@ -361,14 +361,6 @@ const TranscriptInteraction: FC<TranscriptInteractionProps> = memo(props => {
                 return
             }
 
-            if (
-                !editorValue.contextItems.find(contextItem =>
-                    ['repository', 'tree'].includes(contextItem.type)
-                )
-            ) {
-                return
-            }
-
```
This caused prefetching to not run if no @-mentions (repo or dir) were present. This limitation doesn't make sense - we should prefetch even with no mentions.

thenamankumar replied:
It is safe to remove this. This is leftover from the old time when we used `getCodyContext` to retrieve search results which required a repo mention to be present. 

Let me make some more changes here as part of this PR to improve the prefetching logic.

janhartman commented on vscode/src/chat/protocol.ts:
```diff
@@ -226,7 +226,7 @@ interface WebviewEditMessage extends WebviewContextMessage {
     editorState?: unknown | undefined | null
     intent?: ChatMessage['intent'] | undefined | null
     intentScores?: { intent: string; score: number }[] | undefined | null
-    manuallySelectedIntent?: boolean | undefined | null
```
This is the most important change - we track which intent was manually selected properly.

thenamankumar replied:
You actually don't need it because if the `manuallySelectedIntent` is set to `true`, you can use the `intent` field to check which was the manually selected intent. 

As `manuallySelectedIntent of type ChatMessage['intent']` will always be same as `intent`.

camdencheek commented on vscode/src/chat/chat-view/ChatController.ts:
```diff
@@ -698,29 +698,36 @@ export class ChatController implements vscode.Disposable, vscode.WebviewViewProv
+        // If `manuallySelectedIntent` is set, this was a user override, and we should use it.
+        // If `preDetectedIntent` is set, the intent was automatically pre-fetched for the input,
+        // meaning we don't have to fetch it again.
+        const intent = manuallySelectedIntent ? manuallySelectedIntent : preDetectedIntent
```
nit: especially since `manuallySelectedIntent` used to be type boolean, I think this makes it more clear that we only want to use the detected intent when `manuallySelectedIntent` is undefined or null

```suggestion
        const intent = manuallySelectedIntent ?? preDetectedIntent
```

janhartman replied:
Oh nice, I'm still getting used to TS. This is great! :)

thenamankumar reviewed (approved):
@janhartman https://github.com/sourcegraph/cody/pull/6572 is merge. I have rebased this branch on main and resolved the conflicts.

I have pushed a commit to resolve:

> There's still one thing bothering me - the prefetching is too eager because it's based on editor state, meaning if you just click around, select text but make no edits, it'll still trigger.  

Though changing `manuallySelectedIntent` to `ChatMessage['intent']` has no issues as such but I don't think it is required because the `intent` field will tell the manually selected intent whenever `manuallySelectedIntent = true`.

I have tested the latest changes on my side, but please give it another round of testing before merging.

janhartman commented:
> Though changing `manuallySelectedIntent` to `ChatMessage['intent']` has no issues as such but I don't think it is required because the `intent` field will tell the manually selected intent whenever `manuallySelectedIntent = true`.
I think it is. Apart from knowing if the intent was manually selected, you also have to know which one it was. Previously, we were using `detectedIntent` for carrying this info, but that's not OK - you don't know what the model predicted and can't log it. With `manuallySelectedIntent` now storing the info of which intent was selected, `detectedIntent` can store the model-predicted intent.

Thanks for getting the PR in and rebasing this one!

thenamankumar commented:
> > Though changing `manuallySelectedIntent` to `ChatMessage['intent']` has no issues as such but I don't think it is required because the `intent` field will tell the manually selected intent whenever `manuallySelectedIntent = true`.
> 
> I think it is. Apart from knowing if the intent was manually selected, you also have to know which one it was. Previously, we were using `detectedIntent` for carrying this info, but that's not OK - you don't know what the model predicted and can't log it. With `manuallySelectedIntent` now storing the info of which intent was selected, `detectedIntent` can store the model-predicted intent.
> 
> @thenamankumar Thanks for getting the PR in and rebasing this one!

I think this confusion is purely due to the poor choice of variable names in the code by me. The `detectedIntent` in the `ChatController` is actually the intent sendtfrom the webview and not the one from api call inside the ChatController. 

Now the webview can send both pre-fetched intent or manually selected intent as intent. But every time the intent is manually selected the webview sends the manually selected intent only as intent, which is ranamed to `detectedIntent` inside the `ChatController`. 

I am happy to pair to figure this out together.

janhartman commented:
@thenamankumar The situation I originally resolved (and is now broken again sadly) is the following: when correcting the intent, we need to log the intents for telemetry to show that the intent was changed.

First I execute a query and see this for the chat/executed event:
```
"privateMetadata": {
      "detectedIntentScores": {
        "search": 0.9999943971633911,
        "chat": 0.000005611992946796818
      },
      "detectedIntent": "search",
      "userSpecifiedIntent": "auto",
```

Everything here is OK.

Next, I click "Switch to chat" and get this for the next chat/executed event:
```
"privateMetadata": {
      "detectedIntent": "chat",
      "userSpecifiedIntent": "chat",
```

This isn't OK. The detected intent was search, not chat - now, I can't figure out from chat/executed telemetry event that this was indeed a correction. And we're also missing the `detectedIntentScores` field. 

I hope this makes things clearer 😄 This was why I was determined to add the type of manually selected intent as well - we need to pass all of this information to the telemetry to make it usable.

janhartman commented:
@thenamankumar I fixed the bug! But we still have another one: we somehow get onChange events that show the query as being empty. Demo here: https://www.loom.com/share/cfa01b98e7cb4931a338e4359396949a?sid=559950c9-8a32-46bc-94fd-486ae95be9b2 (I just put in a few console.log statements).

janhartman commented:
@thenamankumar I fixed the bug! It was due to assigning the wrong intent to the message.

Furtheremore, I did some more digging and found that we reset intent on focus change which doesn't seem like it's necessary, so I removed it. I tested manually and everything looks great right now.
